src/librustc_data_structures/interner.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::marker::PhantomData;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::hash::BuildHasher;
+use std::mem::{self, size_of};
+use std::ptr::{Unique, NonNull};
+use std::alloc::{Global, Alloc};
+use std::collections::hash_map::RandomState;
+use std::borrow::Borrow;
+use std::fmt;
+
+const ENTRIES_PER_GROUP: usize = 5;
+
+#[repr(align(64), C)]
+pub struct Group {
+    hashes: [u32; ENTRIES_PER_GROUP],
+    size: u32,
+    values: [u64; ENTRIES_PER_GROUP],
+}
+
+impl Group {
+    #[inline(always)]
+    fn search_for_empty(&self) -> Option<usize> {
+        if self.size != ENTRIES_PER_GROUP as u32 {
+            Some(self.size as usize)
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    fn search_with<K, F: FnMut(&K) -> bool>(&self, eq: &mut F, hash: u32) -> Option<(usize, bool)> {
+        for i in 0..ENTRIES_PER_GROUP {
+            let h = unsafe { *self.hashes.get_unchecked(i) };
+            if h == hash && eq(unsafe { mem::transmute(self.values.get_unchecked(i)) }) {
+                return Some((i, false))
+            }
+        }
+        self.search_for_empty().map(|i| (i, true))
+    }
+
+    #[inline(always)]
+    fn set(&mut self, pos: usize, hash: u32, value: u64) {
+        unsafe {
+            *self.hashes.get_unchecked_mut(pos) = hash;
+            *self.values.get_unchecked_mut(pos) = value;
+        }
+    }
+
+    #[inline(always)]
+    fn iter<F: FnMut(u32, u64)>(&self, f: &mut F) {
+        for i in 0..ENTRIES_PER_GROUP {
```

Comment:
If you do `for i in 0 .. self.size` here you can avoid always setting the most significant bit it the hash.

Reply:
LLVM won't unroll the loop in that case

Reply:
Right, but you'll have fewer iterations and you can omit the `h != 0` check. It might be win overall.